File: 0071-simplify-path/0071-simplify-path.cpp
class Solution {
public:
    int getNextSlash(string path,int i){
        int n=path.size();
        for(int k=i;k<n;k++){
            if(path[k]=='/') return k;
        }
        return n;
    }
    void helper(stack<string>& st,vector<string>& v){
        while(st.size()>0){
            v.push_back(st.top());
            st.pop();
        }
    }
    string simplifyPath(string path) {
        int n=path.size();
        stack<string> st;
        int i=1;
        while(i<n){
            int j=getNextSlash(path,i);
            string str=path.substr(i,j-i);
            if(str=="..") {
                if(st.size()>0) st.pop();
            }
            else if(str!="."  && str!="/" && str!="//" && str!=""){
                st.push(str);
            }
            i=j;
            while(i<n && path[i]=='/') i++;
        }
        string ans="";
        vector<string> v;
        helper(st,v);

        n=v.size();
        for(int i=n-1;i>=0;i--){
            ans+='/';
            ans+=v[i];
        }
        if(ans=="") return "/";
        return ans;
    }
};

File: 0071-simplify-path/0071-simplify-path.java
class Solution {
    public String simplifyPath(String path) {
        char[] str=path.toCharArray();
        int n=str.length;

        Stack<String> st=new Stack<>();
        int i=0;
        while(i<n){

            int j=i+1;
            while(j<n && str[j]!='/') j++;

            String sub=path.substring(i+1,j);

            System.out.println(i+"  at this :-> "+sub);

            if (sub.equals("..")) {
            if (!st.isEmpty()) st.pop();
            }
            else if (sub.equals(".") || sub.equals("")) {
                // do nothing
            }
            else {
                st.push(sub);
            }

            
            i=j;
        }
        String ans="";
        String[] temp=new String[st.size()];

        int j=st.size()-1;
        while(st.size()>0){
            temp[j]=st.peek();
            st.pop();
            j--;
        }       
        for(i=0;i<temp.length;i++){
            ans+="/";
            ans+=temp[i];
            System.out.println(temp[i]);
        } 
        if(ans=="") return "/";
        return ans;
    }
}